src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,484 @@
+    auto compare = [&overlaps](const UInt &a, const UInt &b) -> bool
+      {return overlaps[a] > overlaps[b];};
+
+    auto &active = activeCells.getFlatSparse();
+    active.clear();
+    active.reserve(cellsPerInhbitionArea + numDesired * inhibitionAreas );
+
+    for(UInt offset = 0u; offset < activeCells.size; offset += cellsPerInhbitionArea)
+    {
+      // Sort the columns by the amount of overlap.  First make a list of all of
+      // the mini-column indexes.
+      auto activeBegin = active.end();
+      for(UInt i = 0u; i < cellsPerInhbitionArea; i++)
+        active.push_back( i + offset );
+      // Do a partial sort to divide the winners from the losers.  This sort is
+      // faster than a regular sort because it stops after it partitions the
+      // elements about the Nth element, with all elements on their correct side of
+      // the Nth element.
+      std::nth_element(
+        activeBegin,
+        activeBegin + numDesired,
+        active.end(),
+        compare);
+      // Remove the columns which lost the competition.
+      active.resize( active.size() - (cellsPerInhbitionArea - numDesired) );
+      // Finish sorting the winner columns by their overlap.
+      // std::sort(activeBegin, active.end(), compare);
+      // Remove sub-threshold winners
+      // while( active.size() > offset &&
+      //        overlaps[active.back()] < stimulusThreshold_)
+      //     active.pop_back();
+    }
+    activeCells.setFlatSparse( active );
+  }
+
+
+  void learnProximalDendrites( SDR &proximalInputActive,
+                               SDR &proximalInputLearning,
+                               SDR &active ) {
+    SDR AF_SDR( AF->dimensions );
+    auto &activeSegments = AF_SDR.getSparse();
+    for(const auto &cell : active.getFlatSparse())
+    {
+      // Adapt Proximal Segments
+      const auto maxSegment = proximalMaxSegment_[cell];
+      proximalConnections.adaptSegment(maxSegment, proximalInputActive,
+                                       proximalIncrement, proximalDecrement);
+      // connections_.raisePermanencesToThreshold(
+      //                             cell, synPermConnected_, stimulusThreshold_);
+
+      activeSegments[0].push_back(cell);
+      activeSegments[1].push_back(maxSegment);
+    }
+    // TODO: Grow new synapses from the learning inputs?
+
+    AF_SDR.setSparse( activeSegments );
+    AF->addData( AF_SDR );
+  }
+
+
+  void learnDistalDendrites( SDR &activeCells ) {
+
+    // Adapt Predicted Active Cells
+    // TODO
+
+    // Grow Unpredicted Active Cells
+    // TODO
+
+    // Punish Predicted Inactive Cells
+    // TODO
+  }
+
+
+  Real initProximalPermanence(Real connectedPct = 0.5f) {
+    if( rng_.getReal64() <= connectedPct )
+        return proximalSynapseThreshold +
+          (1.0f - proximalSynapseThreshold) * rng_.getReal64();
+      else
+        return proximalSynapseThreshold * rng_.getReal64();
+  }
+};
+
+
+class DefaultTopology : public Topology_t
```

Comment:
later, make Topology a separate file

Reply:
I added this to the check-list at the top of the page, marking this as resolved for now, will fix later.